Python/Product/Analysis/LanguageServer/VolatileCounter.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    sealed class VolatileCounter {
+        private int _count;
+        private TaskCompletionSource<int> _waitForChange;
+        private TaskCompletionSource<object> _waitForZero;
+
+        public int Count => Volatile.Read(ref _count);
+
+        public bool IsZero => Count == 0;
+
+        private sealed class DecrementOnDispose : IDisposable {
+            public VolatileCounter _self;
+            public void Dispose() => _self.Decrement();
```

Comment:
You can ensure that `Decrement` is called only once: `public void Dispose() => Interlocked.Exchange(ref _self, null)?.Decrement();`